env/io_posix.cc

```diff
@@ -945,20 +987,30 @@ Status PosixWritableFile::Allocate(uint64_t offset, uint64_t len) {
 }
 #endif
 
-#ifdef ROCKSDB_RANGESYNC_PRESENT
 Status PosixWritableFile::RangeSync(uint64_t offset, uint64_t nbytes) {
+  // If `sync_file_range` is not available or not supported by the filesystem
+  // on which the file resides, fall back to `Sync()` (`fdatasync`). While this
+  // upholds the contract of `bytes_per_sync`, it has disadvantages: (1) other
+  // non-Posix `Env`s do not have this behavior yet; and (2) unlike
+  // `sync_file_range`, `fdatasync` can sync metadata, increasing write-amp.
+#ifdef ROCKSDB_RANGESYNC_PRESENT
+  if (!sync_file_range_supported_) {
+    return Sync();
+  }
   assert(offset <= std::numeric_limits<off_t>::max());
   assert(nbytes <= std::numeric_limits<off_t>::max());
-  if (sync_file_range(fd_, static_cast<off_t>(offset),
-      static_cast<off_t>(nbytes), SYNC_FILE_RANGE_WRITE) == 0) {
+  if (sync_file_range(fd_, 0, static_cast<off_t>(offset + nbytes),
+                      SYNC_FILE_RANGE_WAIT_BEFORE | SYNC_FILE_RANGE_WRITE) == 0) {
```

Comment:
Are you sure we want to do `SYNC_FILE_RANGE_WAIT_BEFORE`? This will block the write.
The mechanism users supposed to use to guarantee WAL is persistent after a while is DB::SyncWAL(). Is there a reason we want to make sure bytes_per_sync to serve the same purpose too?

Reply:
I don't particularly want to make `wal_bytes_per_sync` serve the same purpose, although I do realize this PR has that side-effect. We have no intention to rely on that behavior.

The point of this PR is to prevent the situation where a file (usually SST, but perhaps WAL as well) is generated fast enough that the async writeback submissions fall behind. When that occurs, upon finishing writing a WAL or SST, the final `fdatasync` involves writing back a huge amount of data. That goes against the expectation a user probably has when they set `bytes_per_sync` or `wal_bytes_per_sync`.

Reply:
But stalling writes waiting for syncing because of bytes_per_sync or wal_bytes_per_sync is a radical behavior change. If you want to introduce another option, I'm fine with that, but the way it is here will create a big surprise to users.

Reply:
Got it. I added a new option `strict_bytes_per_sync` so users can opt-in.